executor/executor.go

```diff
@@ -98,36 +100,85 @@ func (e *TableScanExec) Fields() []*ast.ResultField {
 
 // Next implements Execution Next interface.
 func (e *TableScanExec) Next() (*Row, error) {
-	if e.iter == nil {
-		txn, err := e.ctx.GetTxn(false)
+	for {
+		if e.cursor >= len(e.ranges) {
+			return nil, nil
+		}
+		ran := e.ranges[e.cursor]
+		if e.seekHandle < ran.LowVal {
+			e.seekHandle = ran.LowVal
+		}
+		if e.seekHandle > ran.HighVal {
+			e.cursor++
+			continue
+		}
+		rowKey, err := e.seek()
+		if err != nil || rowKey == nil {
+			return nil, errors.Trace(err)
+		}
+		handle, err := tables.DecodeRecordKeyHandle(rowKey)
 		if err != nil {
 			return nil, errors.Trace(err)
 		}
-		e.iter, err = txn.Seek(e.t.FirstKey())
+		if handle > ran.HighVal {
+			// The handle is out of the current range, but may be in next range.
+			// We try to iterate to the range that contains the handle, so we
+			// don't need to seek again.
+			inRange := e.iterateToRangeContains(handle)
+			if !inRange {
+				continue
+			}
+		}
+		row, err := e.getRow(handle, rowKey)
 		if err != nil {
 			return nil, errors.Trace(err)
 		}
+		e.seekHandle = handle + 1
```

Comment:
Seems now it is OK here, but later we will remove next until method.
Actually in inspectkv, we have already used `handle + 1` as a new seek.